drivers: flash: nrf_qspi_nor: various fixes

pabigot opened the pull request:
The check for small transfers inadvertently allowed a transfer of zero bytes, which should be an error (invalid parameter).

Also ensure the buffer is properly aligned.

Also clean up the lock/unlock idiom to make it easier to support transfers that require multiple HAL calls in a way that doesn't risk interleaving operations from multiple threads.

All this in support of nrfconnect/sdk-zephyr#305

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: flash: nrf_qspi_nor: reject zero length transfers

The check for small transfers inadvertently allowed a transfer of zero
bytes, which should be an error (invalid parameter).

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: flash: nrf_qspi_nor: ensure source buffer is aligned

The stack buffer used to pad small reads must still be aligned to a
word address as the underlying driver uses DMA transfers with that
requirement.  Don't assume the compiler will naturally align it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

de-nordic commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -493,14 +493,14 @@ static int qspi_nor_read(struct device *dev, off_t addr, void *dest,
 
 	/* read size must be non-zero multiple of 4 bytes */
-	if (size < 4U) {
+	if ((size > 0) && (size < 4U)) {
```
Since you are digging here anyway, maybe just gather all the input parameter checks (`addr`, `dest`, `size`)that end up in `-EINVAL` to line 498?

pabigot replied:
The condition becomes:
```
	if ((dest == NULL)
	    || (size == 0)
	    || ((size > 4U) && ((size % 4U) != 0))
	    || ((addr % 4U) != 0)
	    || (addr >= params->size)
	    || (addr < 0)
	    || (size > params->size)
	    || (((addr) + size) > params->size)) {
		return -EINVAL;
	}
```
assuming we don't eliminate the redundant checks but do eliminate the error message for address violations (why is that more important to diagnose than others?).

I can see some merit in that approach, but also merit for checking each parameter separately where we can document the reasons for the check and (where appropriate) issue specific diagnostics.

At this point in the release cycle I prefer to do minimal, more clearly-correct changes that retain consistency with the other functions code.

de-nordic replied:
I was rather obvious checks in mind, like parameter sanity check, these you do not log error for, just return.
```
	if ((dest == NULL)
	    || (size == 0)
	    || ((size > 4U) && ((size % 4U) != 0))
	    || ((addr % 4U) != 0)
}
```
This way we would not do anything that would get discarded anyway on next check (I know that there is not much done here).

> I can see some merit in that approach, but also merit for checking each parameter separately where we can document the reasons for the check and (where appropriate) issue specific diagnostics.

I get your point.

>At this point in the release cycle I prefer to do minimal, more clearly-correct changes that retain consistency with the other functions.

Have not thought about this, my bad.

sigvartmh reviewed (dismissed):
I'm okei with how it is. Early return is always good, but good error messages is also important.

de-nordic reviewed (dismissed):
I think it is OK.

pabigot committed:
drivers: flash: nrf_qspi_nor: clean up lock/unlock idioms

Having a completion wait function release a lock internally only when
the operation fails is confusing.  Remove that feature, and make the
lock and unlock operations explicit and paired.

This makes it much more clear how to properly handle transactions that
require multiple calls to the HAL.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
IMO the addition of the commit that changes the locking regimen invalidated the previous approvals, so @sigvartmh @de-nordic please look again.  It's that change that enables nrfconnect/sdk-zephyr#305 to be safe in the face of multiple thread access by adding the pieces in https://github.com/pabigot/fw-nrfconnect-zephyr/commits/pr/305.

de-nordic commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -319,14 +316,14 @@ static int qspi_erase(struct device *dev, u32_t addr, u32_t size)
 
-	while (size) {
+	qspi_lock(dev);
+	while ((rv == 0) && (size > 0)) {
```
Shouldn't we check if `size` is multiply of full sectors? Otherwise it seems that loop will erase everything it can in sectors/blocks  and then  return with error for the rest of requested range if it has not been fully aligned.

pabigot replied:
Probably, but that should be done in another PR as this is the same loop used in other drivers.

de-nordic commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -353,11 +350,13 @@ static int qspi_erase(struct device *dev, u32_t addr, u32_t size)
 	}
 
-	return 0;
+	qspi_unlock(dev);
```
There will be double `qspi_unlock` when erase size is not sector aligned, because `res` will be set to `!NRFX_SUCCESS`,
and `qspi_wait_for_completion` will be invoked with that `res`, so it will go into `qspi_unlock` path.

pabigot replied:
No: there is no `qspi_unlock()` path in `qspi_wait_for_completion()` anymore.

de-nordic replied:
Sorry got old codebase.

de-nordic reviewed (approved):
I think it is ok.

sigvartmh reviewed (approved):
Looks okei to me and worked with my use-case
